### PR TITLE
fix: avoid silent panic in the main thread loop

### PR DESCRIPTION
**Pull request by Soptq**
Issue Number: #561 

## Type

- [ ] feature
- [x] bugfix
- [ ] dev (no functional changes, no API changes)
- [x] fmt (formatting, renaming)
- [ ] build
- [ ] docs
- [ ] testing

## Description

This PR avoids silent panic in the main thread loop. Moreover, three config parameters (`timeout_sec`, `retry_limit`, `retry_secs`) are added to allow retrying when fetching the state root and the block number failed.

## Breaking changes?

- [ ] yes
- [x] no

### CODE REVIEW

File: crates/core/src/client.rs [hunk 1]
```diff
@@ -152,15 +167,31 @@ impl BeerusClient {
...
-                let sn_block_num = sn_state_block_number_inner(&l1_client, core_contract_addr).await.unwrap();
+                let sn_root = wrapped_retry_sn_state_root_inner(
+                    &l1_client,
+                    core_contract_addr,
+                    config.timeout_secs,
+                    config.retry_limit,
+                    config.retry_secs,
+                )
+                .await
+                .unwrap();
```
Comment (tonypony220) [inline crates/core/src/client.rs:178]: the main idea was to remove `unwrap()` and thus prevent the child thread from panicking

File: crates/core/src/client.rs [hunk 2]
```diff
@@ -279,6 +311,34 @@ async fn sn_state_root_inner(l1_client: &Client<impl Database>, contract_addr: A
     FieldElement::from_byte_slice_be(&starknet_root).map_err(|e| eyre!(e))
 }
 
+async fn wrapped_retry_sn_state_root_inner(
```
Comment (tonypony220) [inline crates/core/src/client.rs:314]: why should we have impl`wrapped_retry_sn_state_root_inner` and `wrapped_retry_sn_state_block_number_inner` separeted?

File: crates/core/src/client.rs [hunk 3]
```diff
@@ -89,6 +89,18 @@ pub struct BeerusClient {
     /// Interval at which beerus will check if the L1 State Root
     /// has been updated.
     pub poll_secs: u64,
+    /// Timeout Seconds
+    ///
+    /// Timeout for fetching L1/L2 state root
+    pub timeout_secs: u64,
```
Comment (tonypony220) [inline crates/core/src/client.rs:95]: Imho, for the current style, hard-coding values within a some resolving scope seem to be sufficient without complicating the configuration

File: crates/core/src/client.rs (hunk 1 above)
Comment (Soptq) [reply crates/core/src/client.rs:178]: Thanks @tonypony220 for your time! I was thinking of not avoiding `unwrap()` but avoiding silent panic in the thread which users are not aware of, as #561 illustrated. So `unwrap` still exists in the main loop, but in `wrapped_retry_*` functions it will catch errors and retry several times. When `retries` reaches the limit, it will return `Err` and make the thread panic and quit, and eventually make the main process quit since `panic = 'abort'` is added. What do you think, I can make the main loop completely ignore the errors, or retry indefinitely.

Comment (Soptq) [reply crates/core/src/client.rs]: i just refactored the `wrapped_retry` function so that the `unwrap()` can be resolved and removed. Now errors are handled in the `wrapped_retry` function, and it either returns the value or shuts the main process down. Moreover, parameters are set in `wrapped_retry` function so we can easily achieve indefinite retrying (namely, never panic).

File: crates/core/src/client.rs (hunk 2 above)
Comment (Soptq) [reply crates/core/src/client.rs]: Fixed. They are merged into the `wrapped_retry` function. I will resolve this conversation but feel free to reopen it ;)

File: crates/core/src/client.rs (hunk 3 above)
Comment (Soptq) [reply crates/core/src/client.rs]: Fixed. They are being hard-coded now in the `wrapped_retry` function. I will resolve this conversation but feel free to reopen it ;)

File: Cargo.toml [hunk 4]
```diff
@@ -3,6 +3,12 @@ members = ["crates/core", "crates/rpc", "crates/cli"]
 exclude = ["examples"]
 resolver = "2"
 
+[profile.release]
+panic = 'abort'
+
+[profile.dev]
+panic = 'abort'
```
[the remark this answers is not in the document]
Comment (Soptq) [reply Cargo.toml:10]: I'm not sure I understand, it seems that we don't need to await `thread::sleep`?

File: crates/core/src/client.rs [hunk 5]
```diff
@@ -290,3 +293,37 @@ async fn sn_state_block_number_inner(
...
+        match future::timeout(time::Duration::from_secs(timeout_secs), func(l1_client, core_contract_addr)).await {
+            Ok(Ok(ret)) => break ret,
+            Ok(Err(err)) => {
+                warn! {"Error for {} times: {}", retries, err}
+            }
+            Err(err) => {
+                warn! {"Timeout for {} times: {}", retries, err}
+            }
+        }
+        thread::sleep(time::Duration::from_secs(retry_secs));
```
[the remark this answers is not in the document]
Comment (Soptq) [reply crates/core/src/client.rs:327]: I'm not sure I understand, it seems we don't need to await `thread::sleep`?

File: crates/core/src/client.rs [hunk 6]
```diff
@@ -152,15 +154,15 @@ impl BeerusClient {
...
-                let sn_block_num = sn_state_block_number_inner(&l1_client, core_contract_addr).await.unwrap();
+                let sn_root = wrapped_retry(sn_state_root_inner, &l1_client, core_contract_addr).await;
+                let sn_block_num = wrapped_retry(sn_state_block_number_inner, &l1_client, core_contract_addr).await;
                 let local_block_num = node.read().await.l1_block_num;
 
                 if local_block_num < sn_block_num {
                     // TODO: Issue #550 - feat: sync from proven root
-                    match l2_client.get_block_with_tx_hashes(BlockId::Tag(SnBlockTag::Latest)).await.unwrap() {
+                    match l2_client.get_block_with_tx_hashes(BlockId::Tag(SnBlockTag::Latest)).await {
                         // TODO: handle unwrap()
```
[the remark this answers is not in the document]
Comment (Soptq) [reply crates/core/src/client.rs:164]: Fixed ;)

File: crates/core/src/client.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (pierre-l) [reply crates/core/src/client.rs:327]: `std::thread::sleep` is synchronous and will keep the future runtime worker for as long as needed. This may lead to worker exhaustion, preventing other futures from being executed concurrently.
You'll probably want `async_std::task::sleep` here, this way the runtime worker will be able to pick up some other work while this future is sleeping.

[the remark this answers is not in the document]
Comment (Soptq) [reply crates/core/src/client.rs]: Oh, got it! Thanks for helping, this is now fixed ;)

File: Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Soptq) [reply Cargo.toml:19]: Thanks for the advice, I have added a comment to explain the motivation to set `panic` to `abort`.

**Comment by Soptq** [comment]
Thanks @pierre-l for the time to review my PR! All requested changes are resolved now, feel free to reopen them if they can be further improved ;)

File: crates/core/src/client.rs [hunk 7]
```diff
@@ -290,3 +292,37 @@ async fn sn_state_block_number_inner(
 
     Ok(U256::from_big_endian(&sn_block_num).as_u64())
 }
+
+async fn wrapped_retry<'a, Fut, T, E, D>(
```
[the remark this answers is not in the document]
Comment (Soptq) [reply crates/core/src/client.rs:296]: Thanks for the time reviewing. 

This crate does not timeout the async function (as mentioned in #561 ) and expects a simple`Result<T>` for retrying. However, when timing the async function (i.e., `future::timeout(duration, async_fn())`), it will return Result<Result<T, CodeError>, TimeoutError>. Thus, when the function is executed in time but raises `CodeError`, additional handling is required. Consequently, considering implementing our custom retry function is not hard, I think we don't need to import and use this crate. What do you think?

File: Cargo.toml [hunk 8]
```diff
@@ -3,6 +3,21 @@ members = ["crates/core", "crates/rpc", "crates/cli"]
...
+# run, hiding the fact that the app has actually crashed.
+# see https://github.com/eigerco/beerus/issues/561 for more details.
+#
+# To avoid this, we abort on panic, which will stop the entire program when a
+# panic occurs, leaving the OS to clean up the mess.
+[profile.release]
+panic = 'abort'
+
+[profile.dev]
+panic = 'abort'
```
[the remark this answers is not in the document]
Comment (Soptq) [reply Cargo.toml:19]: Another approach to achieve this is to use `std::panic::set_hook` to manually exit the process when a panic is encountered. However, this approach is actually the same as the current approach, as the hook is registered globally and applied to all panics. Also, exiting the process in a background thread will also leave the OS to clean up the mess.

Another approach is to use a `AtomicBool` to notify every background threads if some threads encounter panics. They need to query this Bool variable regularly, and when some threads encounter panics, they can get notified and exit gracefully. However, I think this approach introduces too much complexity.

Personally, I think the current approach is the most simple but effective way to achieve what this PR wants to address: avoid silent crash in background threads. Note that errors will normally be handled before final panic (e.g. retrying several times before final panic), so normally panic will never happen. When panic happens, it means something we don't foresee happened. In this case, abort the program is what we want, isn't it? What do you think?

**Comment by Soptq** [comment]
Thanks for the advice! Tests are now added ;)

**Comment by Soptq** [comment]
Thanks for the reply! @sergey-melnychuk 

I have added a detailed response here https://github.com/eigerco/beerus/pull/565#discussion_r1466815005 , once again thanks for your time. The following are additional responses to your latest comment.

> I honestly don't think enforcing non-recoverable error is the right way to handle application-level errors/edge-cases (returning Err should have been enough).

I totally agree with you. However, currently the main problem is that, when `sn_state_root_inner` and `sn_state_block_number_inner` failed, it will not be tried again after a timeout. Instead, it will cause a panic in the client thread, and thereby crashing the client thread. However, by default the crashing in background threads will not disturb other threads including the main thread, making nobody knows the client thread is already dead. Consequently, in practice, users may find their applications seem to work properly, but actually their client threads are dead.

To address this problem, my solution is to simply exit the whole application once the client thread crashes. Thus, users can know their applications are not working. In this way, errors are uncovered in a prominent way instead of a silent way.

> What is the point of retrying specific call, when it will be tried again after a timeout anyway? I honestly don't see why we need retries in this case, with .unwrap() being handled in a more appropriate fashion.

Retrying here is to prevent accidentally exiting the whole application. To be more specific, sometimes `sn_state_root_inner` and `sn_state_block_number_inner` fail due to temporary network issues or other temporary factors. Without retrying, the application will exit very frequently. With retrying, these noises will be cleared, and only serious issues will break the application.

---

To sum up, I think the goal of this PR is not to handle application-level errors/edge-cases, but to uncover silent panics happened in background threads. Without this PR, the application will pretend to be working normally even if its background threads are killed. With this PR, the application will be crashed if any of its background threads crashes.

hope it makes sense.

**Comment by Soptq** [comment]
Thanks @sergey-melnychuk , I have pushed a new implementation that:
1. remove `panic=abort`
2. remove retrying
3. remove `unwrap()`

File: crates/core/src/client.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Soptq) [reply crates/core/src/client.rs]: I will resolve this conversation as a new implementation is pushed that removes retrying. Feel free to reopen this if necessary.

File: Cargo.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (Soptq) [reply Cargo.toml]: I will resolve this conversation as I have removed `panic=abort`. Feel free to reopen it if necessary.
